src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -475,6 +480,24 @@ private bool ConvertTypeToExpression(TypeSyntax type, out ExpressionSyntax expr,
                 case SimpleNameSyntax s:
                     expr = s;
                     return true;
+                case ArrayTypeSyntax array when permitElementAccess:
+                    ExpressionSyntax result = array.ElementType;
+                    foreach (var rank in array.RankSpecifiers)
+                    {
+                        var args = _pool.AllocateSeparated<ArgumentSyntax>();
+                        foreach (GreenNode node in rank.Sizes.GetWithSeparators())
+                        {
+                            if (node.IsToken)
+                                args.AddSeparator(node);
+                            else
+                                args.Add(_syntaxFactory.Argument(null, null, (ExpressionSyntax)node));
+                        }
+                        result = _syntaxFactory.ElementAccessExpression(result,
+                            _syntaxFactory.BracketedArgumentList(rank.OpenBracketToken, args.ToList(), rank.CloseBracketToken));
+                        _pool.Free(args);
+                    }
+                    expr = result;
+                    return true;
```

Comment:
At the moment, this is only for error recovery.

Reply:
I should probably revert this because something like `{ name[0]: p }` will be caught as a recursive pattern and we may not reach here.

Reply:
I agree. I'd think changes to `ConvertTypeToExpression` could be reverted.